engine/src/scriptpt.cpp

```diff
@@ -1310,33 +1309,55 @@ Parse_stat MCScriptPoint::lookup(Script_point t, const LT *&dlt)
 	return PS_NO_MATCH;
 }
 
-bool MCScriptPoint::lookupconstantvalue(const char *& r_value)
+bool MCScriptPoint::lookupconstantintable(uint2& r_position)
 {
-	uint2 high = constant_table_size;
-	uint2 low = 0;
-	int4 cond;
+    uint2 high = constant_table_size;
+    uint2 low = 0;
+    int4 cond;
     
     MCAutoStringRefAsCString t_token;
     t_token . Lock(gettoken_stringref());
     const char *token_cstring = *t_token;
-	while (low < high)
-	{
-		uint2 mid = low + ((high - low) >> 1);
-		cond = MCU_strncasecmp(token_cstring, constant_table[mid].token, token.getlength());
-		if (cond == 0)
-			cond -= constant_table[mid].token[token.getlength()];
-		if (cond < 0)
-			high = mid;
-		else
-			if (cond > 0)
-				low = mid + 1;
-			else
-			{
-                r_value = constant_table[mid] . svalue;
-				return true;
-			}
-	}
-	return false;
+    while (low < high)
+    {
+        uint2 mid = low + ((high - low) >> 1);
+        cond = MCU_strncasecmp(token_cstring, constant_table[mid].token, token.getlength());
+        if (cond == 0)
+        {
+            cond -= constant_table[mid].token[token.getlength()];
+        }
+        
+        if (cond < 0)
+        {
+            high = mid;
+        }
+        else
+        {
+            if (cond > 0)
+            {
+                low = mid + 1;
+            }
+            else
+            {
+                r_position = mid;
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+bool MCScriptPoint::lookupconstantvalue(const char *& r_value)
```

Comment:
Where is lookupconstantvalue used? It seems a bit of an anachronism.

Reply:
Yes I wondered about removing it... here is the code:
```
                // MW-2014-11-06: [[ Bug 3680 ]] If we are in explicit var mode, and the token
                //   is not a string literal or a number, then it must be a constant in the constant
                //   table that is the same as its token.
                if (MCexplicitvariables && type == ST_ID)
                {
                    // If the unquoted literal is a recognised constant and the constant's value
                    // is identical (case-sensitively) to the value, it is fine to make it a literal.
                    const char *t_value;
                    if (sp . lookupconstantvalue(t_value) &&
                        MCStringIsEqualToCString(sp . gettoken_stringref(), t_value, kMCStringOptionCompareExact))
                        type = ST_LIT;
                }
```

So actually this patch reintroduces bug 3680 for `local sVar = true` or `constant kFalse = false`. To fix those we need to tweak this to convert constant values to string in `lookupconstantvalue`.

Reply:
I guess I should just make lookupconstantvalue return a stringref at least

Reply:
Okay so that is slightly irksome - the current behavior of initializers is not ideal...

When explicitVariables is off:
```
local tVar = empty
put tVar
  => "empty" *not* ""
```

This causes a discontinuity if we were to add constant expression initializers:
```
local tVar = empty => "empty"
local tVar = empty & empty => ""
```

I think changing this behavior can be justified as part of introducing constant expressions; but not for just cleaning up how constants work internally.

One way to resolve this is to change lookupconstantvalue so it looks up the constant, and handles the cases which have been broken by the refactor - i.e. true/false/infinity (as they convert as strings to their constant value).